### PR TITLE
data source `azurerm_dns_zone`: make `resource_group_name` required

### PR DESCRIPTION
**Pull request by njuCZ**
according to https://docs.microsoft.com/en-us/azure/dns/dns-zones-records#dns-zones, dns zone name is only unique in resource group. Different resource group could use the same dns zone name. it's not unique global.

So I think it does not make sense to search all resource group and return first found result.

### CODE REVIEW

**Comment by tombuildsstuff** [comment]
hey @njuCZ 

Thanks for this PR.

Taking a look through here - whilst these can be unique per resource group, Terraform should still support looking these up - however Terraform should raise an error if multiple DNS Zones are found with the same name, would you be able to update this PR to fix this instead?

Thanks!

**Comment by njuCZ** [comment]
@tombuildsstuff I have modified this PR, please have a look again.
1. use `listComplete` rather than list all resource group and then ListByResourceGroup
2. when user does not specify `resource_group_name`, reports error if two dns zone have the same name
3. because resource group of dns zone is always lower case, change the test case

File: azurerm/internal/services/dns/dns_zone_data_source.go [hunk 1]
```diff
@@ -109,26 +108,30 @@ func dataSourceArmDnsZoneRead(d *schema.ResourceData, meta interface{}) error {
...
-			return dns.Zone{}, "", fmt.Errorf("Error listing DNS Zones (Resource Group: %s): %+v", resourceGroup, err)
-		}
-
-		for _, z := range zones.Values() {
-			if *z.Name == name {
-				return z, resourceGroup, nil
+	var found *dns.Zone
+	for zonesIterator.NotDone() {
+		zone := zonesIterator.Value()
+		if *zone.Name == name {
```
Comment (tombuildsstuff) [inline azurerm/internal/services/dns/dns_zone_data_source.go:120]: there's a crash here if `zone.Name` is nil

File: azurerm/internal/services/dns/dns_zone_data_source.go [hunk 2]
```diff
@@ -109,26 +108,30 @@ func dataSourceArmDnsZoneRead(d *schema.ResourceData, meta interface{}) error {
...
+			return dns.Zone{}, "", fmt.Errorf("listing DNS Zones: %+v", err)
 		}
 	}
 
-	return dns.Zone{}, "", nil
+	if found == nil {
+		return dns.Zone{}, "", fmt.Errorf("could not find DNS zone with name: %q", name)
+	}
+
+	id, _ := parse.DnsZoneID(*found.ID)
```
Comment (tombuildsstuff) [inline azurerm/internal/services/dns/dns_zone_data_source.go:135]: we should handle this error here

**Comment by njuCZ** [comment]
@tombuildsstuff thanks for your suggestions. I have updated this PR, please have a look again

File: azurerm/internal/services/dns/dns_zone_data_source.go [hunk 3]
```diff
@@ -109,26 +107,33 @@ func dataSourceArmDnsZoneRead(d *schema.ResourceData, meta interface{}) error {
 	return tags.FlattenAndSet(d, resp.Tags)
 }
 
-func findZone(client *dns.ZonesClient, rgClient *resources.GroupsClient, ctx context.Context, name string) (dns.Zone, string, error) {
-	groups, err := rgClient.List(ctx, "", nil)
+func findZone(client *dns.ZonesClient, ctx context.Context, name string) (dns.Zone, string, error) {
```
Comment (tombuildsstuff) [inline azurerm/internal/services/dns/dns_zone_data_source.go:110]: thinking about it this probably should be a pointer:

```suggestion
func findZone(client *dns.ZonesClient, ctx context.Context, name string) (*dns.Zone, string, error) {
```

File: azurerm/internal/services/dns/dns_zone_data_source.go [hunk 4]
```diff
@@ -109,26 +107,33 @@ func dataSourceArmDnsZoneRead(d *schema.ResourceData, meta interface{}) error {
...
+	zonesIterator, err := client.ListComplete(ctx, nil)
 	if err != nil {
-		return dns.Zone{}, "", fmt.Errorf("Error listing Resource Groups: %+v", err)
+		return dns.Zone{}, "", fmt.Errorf("listing DNS Zones: %+v", err)
```
Comment (tombuildsstuff) [inline azurerm/internal/services/dns/dns_zone_data_source.go:113]: that allows this to be nil - ensuring that if we return an error, trying to access the DNS Zone should panic - which'll be caught by tests (and forces us to check errors first):

```suggestion
		return nil, "", fmt.Errorf("listing DNS Zones: %+v", err)
```

File: azurerm/internal/services/dns/dns_zone_data_source.go [hunk 5]
```diff
@@ -109,26 +107,33 @@ func dataSourceArmDnsZoneRead(d *schema.ResourceData, meta interface{}) error {
...
-
-		for _, z := range zones.Values() {
-			if *z.Name == name {
-				return z, resourceGroup, nil
+	var found *dns.Zone
+	for zonesIterator.NotDone() {
+		zone := zonesIterator.Value()
+		if zone.Name != nil && *zone.Name == name {
+			if found != nil {
+				return dns.Zone{}, "", fmt.Errorf("found multiple DNS zones with name %q, please specify the resource group", name)
```
Comment (tombuildsstuff) [inline azurerm/internal/services/dns/dns_zone_data_source.go:121]:
```suggestion
				return nil, "", fmt.Errorf("found multiple DNS zones with name %q, please specify the resource group", name)
```

File: azurerm/internal/services/dns/dns_zone_data_source.go [hunk 6]
```diff
@@ -109,26 +107,33 @@ func dataSourceArmDnsZoneRead(d *schema.ResourceData, meta interface{}) error {
...
 			}
+			found = &zone
 		}
+		if err := zonesIterator.NextWithContext(ctx); err != nil {
+			return dns.Zone{}, "", fmt.Errorf("listing DNS Zones: %+v", err)
```
Comment (tombuildsstuff) [inline azurerm/internal/services/dns/dns_zone_data_source.go:126]:
```suggestion
			return nil, "", fmt.Errorf("listing DNS Zones: %+v", err)
```

File: azurerm/internal/services/dns/dns_zone_data_source.go [hunk 7]
```diff
@@ -109,26 +107,33 @@ func dataSourceArmDnsZoneRead(d *schema.ResourceData, meta interface{}) error {
...
+		}
+	}
+
+	if found == nil || found.ID == nil {
+		return dns.Zone{}, "", fmt.Errorf("could not find DNS zone with name: %q", name)
```
Comment (tombuildsstuff) [inline azurerm/internal/services/dns/dns_zone_data_source.go:131]:
```suggestion
		return nil, "", fmt.Errorf("could not find DNS zone with name: %q", name)
```

File: azurerm/internal/services/dns/dns_zone_data_source.go [hunk 8]
```diff
@@ -109,26 +107,33 @@ func dataSourceArmDnsZoneRead(d *schema.ResourceData, meta interface{}) error {
...
 	}
 
-	return dns.Zone{}, "", nil
+	id, err := parse.DnsZoneID(*found.ID)
+	if err != nil {
+		return dns.Zone{}, "", fmt.Errorf("DNS zone id not valid: %+v", err)
```
Comment (tombuildsstuff) [inline azurerm/internal/services/dns/dns_zone_data_source.go:136]:
```suggestion
		return nil, "", fmt.Errorf("DNS zone id not valid: %+v", err)
```

File: azurerm/internal/services/dns/dns_zone_data_source.go [hunk 9]
```diff
@@ -109,26 +107,33 @@ func dataSourceArmDnsZoneRead(d *schema.ResourceData, meta interface{}) error {
...
+	}
+	return *found, id.ResourceGroup, nil
```
Comment (tombuildsstuff) [inline azurerm/internal/services/dns/dns_zone_data_source.go:138]: this should be de-referenced anyway - so it should be fine:

```suggestion
	return found, id.ResourceGroup, nil
```

File: azurerm/internal/services/dns/dns_zone_data_source.go [hunk 10]
```diff
@@ -77,16 +77,17 @@ func dataSourceArmDnsZoneRead(d *schema.ResourceData, meta interface{}) error {
...
 			return err
 		}
 
-		if resourceGroup == "" {
+		if zone == nil {
 			return fmt.Errorf("Error: DNS Zone %q was not found", name)
 		}
+
+		resp = *zone
 	}
```
Comment (njuCZ) [inline azurerm/internal/services/dns/dns_zone_data_source.go:91]: @tombuildsstuff thanks for your suggestions. I have updated this function to return a pointer. In order to use the same variable `resp` and `resourcegroup`, I made a little update here, please have a check again.

**Comment by njuCZ** [comment]
@tombuildsstuff thanks for your suggestions. I have updated this function to return a pointer. in order to use the same variable `resp` and `resourceGroup`, I made a little change in the read func, please have a check

**Comment by jackofallops** [comment]
Tests passing:
image

**Pull request merged**